extensions/amp-fx-collection/0.1/providers/parallax.js

```diff
@@ -0,0 +1,169 @@
+ * limitations under the License.
+ */
+
+import {
+  installPositionObserverServiceForDoc,
+} from '../../../../src/service/position-observer/position-observer-impl';
+import {
+  PositionObserverFidelity,
+} from '../../../../src/service/position-observer/position-observer-worker';
+import {user} from '../../../../src/log';
+import {getServiceForDoc} from '../../../../src/service';
+import {Services} from '../../../../src/services';
+import {setStyles} from '../../../../src/style';
+
+const FACTOR_ATTR = 'data-parallax-factor';
+
+/**
+ * Provides a parallax visual effect given a parallax factor.
+ *
+ * @implements {../amp-fx-collection.FxProviderInterface}
+ */
+export class ParallaxProvider {
+
+  /**
+   * @param  {!../../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+
+    /** @private @const {!../../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @private @const {!../../../../src/service/viewport/viewport-impl.Viewport} */
+    this.viewport_ = Services.viewportForDoc(ampdoc);
+
+    installPositionObserverServiceForDoc(ampdoc);
+
+    /** @private @const {!../../../../src/service/position-observer/position-observer-impl.PositionObserver} */
+    this.positionObserver_ = getServiceForDoc(ampdoc, 'position-observer');
+  }
+
+  /**
+   * Installs parallax effect on the element
+   * @param {!Element} element
+   */
+  installOn(element) {
+    Services.vsyncFor(this.ampdoc_.win).measure(() => {
+      new ParallaxElement(element, this.viewport_, this.positionObserver_);
+    });
+  }
+}
+
+/**
+ * Encapsulates and tracks an element's linear parallax effect.
+ */
+class ParallaxElement {
+  /**
+   * @param {!Element} element The element to give a parallax effect.
+   * @param {!../../../../src/service/viewport/viewport-impl.Viewport} viewport
+   * @param {!../../../../src/service/position-observer/position-observer-impl.PositionObserver} positionObserver
+   */
+  constructor(element, viewport, positionObserver) {
+
+    /** @private @const {!../../../../src/service/position-observer/position-observer-impl.PositionObserver} */
+    this.positionObserver_ = positionObserver;
+
+    /** @private @const {!../../../../src/service/viewport/viewport-impl.Viewport} */
+    this.viewport_ = viewport;
+
+    /** @const {string} */
+    const factorValue = user().assert(element.getAttribute(FACTOR_ATTR),
+        `${FACTOR_ATTR}=<number> attribute must be provided for: %s`, element);
+
+    /** @private @const {number} */
+    this.factor_ = parseFloat(factorValue);
+
+    user().assert(this.factor_ > 0,
+        `${FACTOR_ATTR} must be a number and greater than 0 for: %s`, element);
+
+    /** @private @const {!Element} */
+    this.element_ = element;
+
+    /** @private {number} */
+    this.adjustedViewportHeight_ = this.getAdjustedViewportHeight_();
+
+    // start observing position of the element.
+    this.observePositionChanges_();
```

Comment:
I notice that some of your constructors in this PR mutate global state, e.g. adding event listeners, installing services. You can mutate global state, but then you have to stub those functions in tests which makes writing tests more difficult. I think that having pure constructors may make testing easier, and just have the consumers of the class call the state-mutating functions like `observePositionChanges()`.

Reply:
done